File: engine/zeromiss/tracker.py
"""Tracker-in-the-loop — radar estimation feeding the seeker.

A real interceptor doesn't get the target's truth; it gets a *noisy* radar return and
must **estimate** the target's state before it can guide on it. This module adds that
layer — estimation and sensor fusion — on top of the guidance core.

Contents
--------
* :func:`radar_measure` — a noisy range/bearing measurement of the target from the
  (known) interceptor position.
* :class:`EKF` — an Extended Kalman Filter tracking the target's
  ``[x, y, vx, vy]`` from those polar measurements (constant-velocity model).
* :class:`IMM` — an Interacting Multiple Model filter blending a *quiet* (low
  process-noise) and a *maneuvering* (high process-noise) EKF, so it tracks a steady
  target tightly **and** a hard-maneuvering one without lag — the textbook win of IMM
  over a single model.
* :class:`Tracker` — the engagement-facing wrapper: ``step()`` predicts every physics
  step and corrects at the radar rate, exposing an estimated target
  :class:`~zeromiss.frames.Kinematic` the guidance law can use.

All randomness flows from the single seeded ``Generator``.
"""

from __future__ import annotations

import math
from dataclasses import dataclass, field

import numpy as np

from .frames import Kinematic, wrap_to_pi


def radar_measure(
    missile: Kinematic, target: Kinematic, rng, sigma_r: float, sigma_beta: float
) -> tuple[float, float]:
    """Return a noisy ``(range, bearing)`` of the target from the interceptor.

    ``range`` [m] and ``bearing`` [rad] (absolute, atan2(dy, dx)) with independent
    Gaussian noise of std ``sigma_r`` and ``sigma_beta``.
    """
    dx = target.x - missile.x
    dy = target.y - missile.y
    r = math.hypot(dx, dy) + rng.normal(0.0, sigma_r)
    beta = math.atan2(dy, dx) + rng.normal(0.0, sigma_beta)
    return r, beta


def _Q_cv(dt: float, q: float) -> np.ndarray:
    """Discrete white-noise-acceleration process covariance for a CV model."""
    dt2 = dt * dt
    dt3 = dt2 * dt
    dt4 = dt2 * dt2
    return q * np.array([
        [dt4 / 4, 0, dt3 / 2, 0],
        [0, dt4 / 4, 0, dt3 / 2],
        [dt3 / 2, 0, dt2, 0],
        [0, dt3 / 2, 0, dt2],
    ])


class EKF:
    """Extended Kalman Filter: target ``[x, y, vx, vy]`` from range/bearing.

    Parameters
    ----------
    q : float
        Process-noise intensity (m^2/s^3). Small => trusts the constant-velocity model;
        large => adapts fast to maneuvers (at the cost of noisier estimates).
    sigma_r, sigma_beta : float
        Measurement noise std (range [m], bearing [rad]).
    """

    def __init__(self, q: float = 50.0, sigma_r: float = 20.0, sigma_beta: float = 5e-3) -> None:
        self.q = q
        self.R = np.diag([sigma_r**2, sigma_beta**2])
        self.x = np.zeros(4)
        self.P = np.eye(4) * 1e6
        self._init = False

    def initialize(self, x: float, y: float, vx: float = 0.0, vy: float = 0.0) -> None:
        self.x = np.array([x, y, vx, vy], dtype=float)
        self.P = np.diag([1e3, 1e3, 1e4, 1e4]).astype(float)
        self._init = True

    def predict(self, dt: float) -> None:
        F = np.array([
            [1, 0, dt, 0],
            [0, 1, 0, dt],
            [0, 0, 1, 0],
            [0, 0, 0, 1],
        ], dtype=float)
        self.x = F @ self.x
        self.P = F @ self.P @ F.T + _Q_cv(dt, self.q)

    def _hH(self, sx: float, sy: float) -> tuple[np.ndarray, np.ndarray]:
        dx = self.x[0] - sx
        dy = self.x[1] - sy
        r2 = dx * dx + dy * dy
        r = math.sqrt(max(r2, 1e-9))
        h = np.array([r, math.atan2(dy, dx)])
        H = np.array([
            [dx / r, dy / r, 0.0, 0.0],
            [-dy / r2, dx / r2, 0.0, 0.0],
        ])
        return h, H

    def update(self, z: tuple[float, float], sx: float, sy: float) -> float:
        """Correct with a measurement; return the measurement likelihood (for IMM)."""
        h, H = self._hH(sx, sy)
        y = np.array([z[0] - h[0], wrap_to_pi(z[1] - h[1])])
        S = H @ self.P @ H.T + self.R
        Sinv = np.linalg.inv(S)
        K = self.P @ H.T @ Sinv
        self.x = self.x + K @ y
        self.P = (np.eye(4) - K @ H) @ self.P
        # Gaussian likelihood of the innovation (for IMM mode-probability update)
        det = max(np.linalg.det(2 * np.pi * S), 1e-12)
        return float(math.exp(-0.5 * y @ Sinv @ y) / math.sqrt(det))

    def estimate(self) -> np.ndarray:
        return self.x.copy()


class IMM:
    """Interacting Multiple Model filter over a quiet and a maneuvering CV-EKF.

    Two models share the same state, so mixing is exact. The Markov transition matrix
    ``Pi`` governs how readily the filter switches between "the target is coasting" and
    "the target is maneuvering" — letting it track both regimes without the lag a single
    model suffers.
    """

    def __init__(
        self,
        q_quiet: float = 5.0,
        q_maneuver: float = 5000.0,
        sigma_r: float = 20.0,
        sigma_beta: float = 5e-3,
        pi_stay: float = 0.95,
    ) -> None:
        self.models = [
            EKF(q_quiet, sigma_r, sigma_beta),
            EKF(q_maneuver, sigma_r, sigma_beta),
        ]
        self.mu = np.array([0.5, 0.5])
        self.Pi = np.array([[pi_stay, 1 - pi_stay], [1 - pi_stay, pi_stay]])

    def initialize(self, x: float, y: float, vx: float = 0.0, vy: float = 0.0) -> None:
        for m in self.models:
            m.initialize(x, y, vx, vy)

    def step(self, dt: float, z: tuple[float, float], sx: float, sy: float) -> None:
        n = len(self.models)
        # --- mixing ---
        cbar = self.Pi.T @ self.mu
        cbar = np.where(cbar < 1e-12, 1e-12, cbar)
        mix = (self.Pi * self.mu[:, None]) / cbar[None, :]  # mix[i,j]
        xs = [m.estimate() for m in self.models]
        Ps = [m.P.copy() for m in self.models]
        for j in range(n):
            x0 = sum(mix[i, j] * xs[i] for i in range(n))
            P0 = np.zeros((4, 4))
            for i in range(n):
                d = (xs[i] - x0).reshape(4, 1)
                P0 += mix[i, j] * (Ps[i] + d @ d.T)
            self.models[j].x = x0
            self.models[j].P = P0
        # --- mode-matched predict + update, collect likelihoods ---
        likelihood = np.zeros(n)
        for j, m in enumerate(self.models):
            m.predict(dt)
            likelihood[j] = m.update(z, sx, sy)
        # --- mode probability update ---
        self.mu = cbar * likelihood
        s = self.mu.sum()
        self.mu = self.mu / s if s > 1e-12 else np.ones(n) / n

    def estimate(self) -> np.ndarray:
        return sum(self.mu[j] * self.models[j].estimate() for j in range(len(self.models)))

    @property
    def maneuver_probability(self) -> float:
        return float(self.mu[1])


@dataclass
class TrackerConfig:
    enabled: bool = False
    model: str = "imm"          # "ekf" | "imm"
    sigma_r_m: float = 20.0     # radar range noise [m]
    sigma_beta_mrad: float = 5.0  # radar bearing noise [mrad]
    radar_hz: float = 50.0      # measurement update rate
    q: float = 50.0             # EKF process-noise intensity
    q_quiet: float = 5.0        # IMM quiet-model process noise
    q_maneuver: float = 5000.0  # IMM maneuvering-model process noise


@dataclass
class Tracker:
    """Engagement-facing wrapper: predict every step, correct at the radar rate."""

    cfg: TrackerConfig
    _filter: object = field(default=None, init=False)
    _t_since_update: float = field(default=math.inf, init=False)
    _started: bool = field(default=False, init=False)

    def __post_init__(self) -> None:
        sigma_beta = self.cfg.sigma_beta_mrad * 1e-3
        if self.cfg.model.lower() == "ekf":
            self._filter = EKF(self.cfg.q, self.cfg.sigma_r_m, sigma_beta)
        else:
            self._filter = IMM(self.cfg.q_quiet, self.cfg.q_maneuver, self.cfg.sigma_r_m, sigma_beta)

    @property
    def update_dt(self) -> float:
        return 1.0 / self.cfg.radar_hz if self.cfg.radar_hz > 0 else 0.0

    def step(self, missile: Kinematic, target: Kinematic, dt: float, rng) -> Kinematic:
        """Advance the tracker one physics step; return the estimated target Kinematic."""
        f = self._filter
        if not self._started:
            f.initialize(target.x, target.y, target.vx, target.vy)  # type: ignore[attr-defined]
            self._started = True
            self._t_since_update = 0.0
            return target

        self._t_since_update += dt
        if self._t_since_update + 1e-12 >= self.update_dt:
            Ts = self._t_since_update
            self._t_since_update = 0.0
            z = radar_measure(missile, target, rng, self.cfg.sigma_r_m, self.cfg.sigma_beta_mrad * 1e-3)
            if isinstance(f, IMM):
                f.step(Ts, z, missile.x, missile.y)
            else:
                f.predict(Ts)
                f.update(z, missile.x, missile.y)

        est = f.estimate()
        speed = math.hypot(est[2], est[3])
        heading = math.atan2(est[3], est[2]) if speed > 1e-6 else target.heading
        return Kinematic(est[0], est[1], speed if speed > 1e-6 else target.speed, heading)

File: web/lib/sim/tracker.ts
/**
 * Tracker-in-the-loop (TS port of `engine/zeromiss/tracker.py`): radar EKF + IMM.
 *
 * Estimates the target's [x, y, vx, vy] from noisy range/bearing measurements. The IMM
 * blends a quiet and a maneuvering EKF so it tracks both a coasting and a hard-turning
 * target without the lag a single model suffers. Used by the /tracker demo page.
 */

import type { RNG } from "./rng";

type Vec4 = [number, number, number, number];
type Mat4 = number[][]; // 4x4

export function radarMeasure(
  mx: number, my: number, tx: number, ty: number, rng: RNG, sigmaR: number, sigmaBeta: number,
): [number, number] {
  const dx = tx - mx;
  const dy = ty - my;
  return [Math.hypot(dx, dy) + rng.normal(0, sigmaR), Math.atan2(dy, dx) + rng.normal(0, sigmaBeta)];
}

function wrapPi(a: number): number {
  return ((a + Math.PI) % (2 * Math.PI) + 2 * Math.PI) % (2 * Math.PI) - Math.PI;
}

function eye4(s = 1): Mat4 {
  return [
    [s, 0, 0, 0],
    [0, s, 0, 0],
    [0, 0, s, 0],
    [0, 0, 0, s],
  ];
}

function matAdd(A: Mat4, B: Mat4): Mat4 {
  return A.map((row, i) => row.map((v, j) => v + B[i][j]));
}

function mat4mul(A: Mat4, B: Mat4): Mat4 {
  const C: Mat4 = eye4(0);
  for (let i = 0; i < 4; i++)
    for (let j = 0; j < 4; j++) {
      let s = 0;
      for (let k = 0; k < 4; k++) s += A[i][k] * B[k][j];
      C[i][j] = s;
    }
  return C;
}

function Qcv(dt: number, q: number): Mat4 {
  const dt2 = dt * dt;
  const dt3 = dt2 * dt;
  const dt4 = dt2 * dt2;
  return [
    [(q * dt4) / 4, 0, (q * dt3) / 2, 0],
    [0, (q * dt4) / 4, 0, (q * dt3) / 2],
    [(q * dt3) / 2, 0, q * dt2, 0],
    [0, (q * dt3) / 2, 0, q * dt2],
  ];
}

export class EKF {
  x: Vec4 = [0, 0, 0, 0];
  P: Mat4 = eye4(1e6);
  constructor(public q = 50, public sigmaR = 20, public sigmaBeta = 5e-3) {}

  initialize(x: number, y: number, vx = 0, vy = 0): void {
    this.x = [x, y, vx, vy];
    this.P = [
      [1e3, 0, 0, 0],
      [0, 1e3, 0, 0],
      [0, 0, 1e4, 0],
      [0, 0, 0, 1e4],
    ];
  }

  predict(dt: number): void {
    const F: Mat4 = [
      [1, 0, dt, 0],
      [0, 1, 0, dt],
      [0, 0, 1, 0],
      [0, 0, 0, 1],
    ];
    this.x = [
      this.x[0] + dt * this.x[2],
      this.x[1] + dt * this.x[3],
      this.x[2],
      this.x[3],
    ];
    this.P = matAdd(mat4mul(mat4mul(F, this.P), transpose(F)), Qcv(dt, this.q));
  }

  /** Correct with z=[r,beta] from sensor (sx,sy); returns the innovation likelihood. */
  update(z: [number, number], sx: number, sy: number): number {
    const dx = this.x[0] - sx;
    const dy = this.x[1] - sy;
    const r2 = Math.max(dx * dx + dy * dy, 1e-9);
    const r = Math.sqrt(r2);
    const h: [number, number] = [r, Math.atan2(dy, dx)];
    // H is 2x4
    const H = [
      [dx / r, dy / r, 0, 0],
      [-dy / r2, dx / r2, 0, 0],
    ];
    const y: [number, number] = [z[0] - h[0], wrapPi(z[1] - h[1])];
    // S = H P H^T + R   (2x2)
    const HP = mul(H, this.P); // 2x4
    const S = add2(mul2(HP, transpose(H) as number[][]), [
      [this.sigmaR ** 2, 0],
      [0, this.sigmaBeta ** 2],
    ]);
    const Sinv = inv2(S);
    // K = P H^T S^-1  (4x2)
    const PHt = mul(this.P, transpose(H) as number[][]); // 4x2
    const K = mul(PHt, Sinv); // 4x2
    // x += K y
    for (let i = 0; i < 4; i++) this.x[i] += K[i][0] * y[0] + K[i][1] * y[1];
    // P = (I - K H) P
    const KH = mul(K, H); // 4x4
    const ImKH: Mat4 = eye4(1).map((row, i) => row.map((v, j) => v - KH[i][j]));
    this.P = mat4mul(ImKH, this.P);
    // likelihood
    const det = Math.max(2 * Math.PI * Math.sqrt(S[0][0] * S[1][1] - S[0][1] * S[1][0]), 1e-12);
    const m = y[0] * (Sinv[0][0] * y[0] + Sinv[0][1] * y[1]) + y[1] * (Sinv[1][0] * y[0] + Sinv[1][1] * y[1]);
    return Math.exp(-0.5 * m) / det;
  }

  estimate(): Vec4 {
    return [...this.x] as Vec4;
  }
}

export class IMM {
  models: EKF[];
  mu = [0.5, 0.5];
  Pi: number[][];
  constructor(qQuiet = 5, qManeuver = 5000, sigmaR = 20, sigmaBeta = 5e-3, piStay = 0.95) {
    this.models = [new EKF(qQuiet, sigmaR, sigmaBeta), new EKF(qManeuver, sigmaR, sigmaBeta)];
    this.Pi = [
      [piStay, 1 - piStay],
      [1 - piStay, piStay],
    ];
  }
  initialize(x: number, y: number, vx = 0, vy = 0): void {
    for (const m of this.models) m.initialize(x, y, vx, vy);
  }
  step(dt: number, z: [number, number], sx: number, sy: number): void {
    const n = 2;
    const cbar = [0, 0];
    for (let j = 0; j < n; j++) for (let i = 0; i < n; i++) cbar[j] += this.Pi[i][j] * this.mu[i];
    for (let j = 0; j < n; j++) cbar[j] = Math.max(cbar[j], 1e-12);
    const xs = this.models.map((m) => m.estimate());
    const Ps = this.models.map((m) => m.P.map((r) => [...r]));
    for (let j = 0; j < n; j++) {
      const x0: Vec4 = [0, 0, 0, 0];
      for (let i = 0; i < n; i++) {
        const w = (this.Pi[i][j] * this.mu[i]) / cbar[j];
        for (let k = 0; k < 4; k++) x0[k] += w * xs[i][k];
      }
      const P0: Mat4 = eye4(0);
      for (let i = 0; i < n; i++) {
        const w = (this.Pi[i][j] * this.mu[i]) / cbar[j];
        const d = [xs[i][0] - x0[0], xs[i][1] - x0[1], xs[i][2] - x0[2], xs[i][3] - x0[3]];
        for (let a = 0; a < 4; a++) for (let b = 0; b < 4; b++) P0[a][b] += w * (Ps[i][a][b] + d[a] * d[b]);
      }
      this.models[j].x = x0;
      this.models[j].P = P0;
    }
    const L = [0, 0];
    for (let j = 0; j < n; j++) {
      this.models[j].predict(dt);
      L[j] = this.models[j].update(z, sx, sy);
    }
    let s = 0;
    for (let j = 0; j < n; j++) {
      this.mu[j] = cbar[j] * L[j];
      s += this.mu[j];
    }
    for (let j = 0; j < n; j++) this.mu[j] = s > 1e-12 ? this.mu[j] / s : 1 / n;
  }
  estimate(): Vec4 {
    const e: Vec4 = [0, 0, 0, 0];
    for (let j = 0; j < this.models.length; j++) {
      const x = this.models[j].estimate();
      for (let k = 0; k < 4; k++) e[k] += this.mu[j] * x[k];
    }
    return e;
  }
  get maneuverProbability(): number {
    return this.mu[1];
  }
}

// --- tiny generic matrix helpers (rows x cols as number[][]) ---
function transpose(A: number[][]): number[][] {
  return A[0].map((_, j) => A.map((row) => row[j]));
}
function mul(A: number[][], B: number[][]): number[][] {
  const r = A.length;
  const c = B[0].length;
  const k0 = B.length;
  const C = Array.from({ length: r }, () => new Array(c).fill(0));
  for (let i = 0; i < r; i++) for (let j = 0; j < c; j++) { let s = 0; for (let k = 0; k < k0; k++) s += A[i][k] * B[k][j]; C[i][j] = s; }
  return C;
}
function mul2(A: number[][], B: number[][]): number[][] {
  return mul(A, B);
}
function add2(A: number[][], B: number[][]): number[][] {
  return A.map((row, i) => row.map((v, j) => v + B[i][j]));
}
function inv2(S: number[][]): number[][] {
  const det = S[0][0] * S[1][1] - S[0][1] * S[1][0];
  const d = Math.abs(det) < 1e-12 ? 1e-12 : det;
  return [
    [S[1][1] / d, -S[0][1] / d],
    [-S[1][0] / d, S[0][0] / d],
  ];
}
